clicommand/pipeline_eval.go

```diff
@@ -0,0 +1,217 @@
+		done := HandleGlobalFlags(l, cfg)
+		defer done()
+
+		// Find the pipeline file either from STDIN or the first
+		// argument
+		var input []byte
+		var filename string
+
+		if cfg.FilePath != "" {
+			l.Info("Reading pipeline config from \"%s\"", cfg.FilePath)
+
+			filename = filepath.Base(cfg.FilePath)
+			input, err = os.ReadFile(cfg.FilePath)
+			if err != nil {
+				l.Fatal("Failed to read file: %s", err)
+			}
+		} else if stdin.IsReadable() {
+			l.Info("Reading pipeline config from STDIN")
+
+			// Actually read the file from STDIN
+			input, err = io.ReadAll(os.Stdin)
+			if err != nil {
+				l.Fatal("Failed to read from STDIN: %s", err)
+			}
+		} else {
+			l.Info("Searching for pipeline config...")
+
+			paths := []string{
+				"buildkite.js",
+				filepath.FromSlash(".buildkite/buildkite.js"),
+				filepath.FromSlash("buildkite/buildkite.js"),
+			}
+
+			// Collect all the files that exist
+			exists := []string{}
+			for _, path := range paths {
+				if _, err := os.Stat(path); err == nil {
+					exists = append(exists, path)
+				}
+			}
+
+			// If more than 1 of the config files exist, throw an
+			// error. There can only be one!!
+			if len(exists) > 1 {
+				l.Fatal("Found multiple configuration files: %s. Please only have 1 configuration file present.", strings.Join(exists, ", "))
+			} else if len(exists) == 0 {
+				l.Fatal("Could not find a default pipeline configuration file. See `buildkite-agent pipeline upload --help` for more information.")
+			}
+
+			found := exists[0]
+
+			l.Info("Found config file \"%s\"", found)
+
+			// Read the default file
+			filename = path.Base(found)
+			input, err = os.ReadFile(found)
+			if err != nil {
+				l.Fatal("Failed to read file \"%s\" (%s)", found, err)
+			}
+		}
+
+		if err := evalJS(filename, input, c.App.Writer); err != nil {
+			panic(err)
+		}
+		return nil
+	},
+}
+
+func evalJS(filename string, input []byte, output io.Writer) error {
+	runtime := goja.New()
+	runtime.SetFieldNameMapper(goja.TagFieldNameMapper("json", true))
+
+	// Add support for require() CommonJS modules.
+	// require("buildkite/*") is handled by embedded resources/node_modules/buildkite/* filesystem.
+	// Other paths are loaded from the host filesystem.
+	registry := require.NewRegistry(
+		require.WithLoader(func(name string) ([]byte, error) {
+			if !strings.HasPrefix(name, "node_modules/buildkite/") {
+				return require.DefaultSourceLoader(name)
+			}
+			res := resources.FS
+			data, err := res.ReadFile(name)
+			if errors.Is(err, fs.ErrNotExist) {
+				return nil, require.ModuleFileDoesNotExistError
+			} else if err != nil {
+				return nil, err
+			}
+			return data, nil
+		}),
+	)
+	registry.Enable(runtime)
+
+	// Add basic utilities
+	console.Enable(runtime) // console.log()
+	process.Enable(runtime) // process.env()
+
+	// provide plugin() as a native module (implemented in Go)
+	registry.RegisterNativeModule("buildkite/plugin", func(runtime *goja.Runtime, module *goja.Object) {
```

Comment:
Ohhh, this is how you did it. I can imagine expanding on the API would be a bit of a pain if it's all in Golang. Presumably we could `RunScript` with some existing gear in the VM first?

Reply:
@keithpitt Yeah this `registry.RegisterNativeModule("buildkite/plugin", …` is just a proof-of-concept of JS functions implemented in Go, but I think you'd want a very good reason to do that; like interacting with existing/complex Go code/libs.

The `require.WithLoader(func(name string) ([]byte, error) { …` ~20 lines further up loads `require("buildkite/*")` straight out of `resources/node_modules/buildkite/*.js` files which get embedded into the `buildkite-agent` binary at compile time and exposed via a virtual read-only filesystem.